text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+#### Duplicate Data
+
+In edge cases (e.g. on reconnections, network interruptions, etc) the client has no way of knowing if recently sent data was delivered if no acknowledgement was received yet. The client will typically choose to re-send such data to guarantee delivery, which may result in duplicate data on the server side. This is a deliberate choice and is considered to be the right tradeoff for telemetry data.
+
+### Client Mode
+
+Clients that want to avoid complex implementations can choose to implement Synchronous mode only.
+
+## Future Versions and Interoperability
+
+OTLP will evolve and change over time. Future versions of OTLP must be designed and implemented in a way that ensures that clients and servers that implement different versions of OTLP can interoperate and exchange telemetry data. Old clients must be able to talk to new servers and vice versa. If new versions of OTLP introduce new functionality that cannot be understood and supported by nodes implementing the old versions of OTLP the protocol must regress to the lowest common denominator from functional perspective.
+
+When possible the interoperability SHOULD be ensured between all versions of OTLP that are not declared obsolete.
+
+OTLP does not use explicit protocol version numbering. OTLP's interoperability of clients and servers of different versions is based on the following concepts:
+
+1. OTLP (current and future versions) defines a set of capabilities, some of which are mandatory, others are optional. Clients and servers must implement mandatory capabilities and can choose implement only a subset of optional capabilities.
+
+2. For minor changes to the protocol future versions and extension of OTLP are encouraged to use the ability of Protocol Buffers to evolve message schema in backwards compatible manner. Newer versions of OTLP may add new fields to messages that will be ignored by clients and servers that do not understand these fields. In many cases careful design of such schema changes and correct choice of default values for new fields is enough to ensure interoperability of different versions without nodes explicitly detecting that their peer node has different capabilities.
+
+3. More significant changes must be explicitly defined as new optional capabilities. Such capabilities SHOULD be discovered by client and server implementations during `Hello` handshake. The capabilities SHOULD be described by the payload contained in `Hello` request and response. The mandatory capabilities defined by this specification are implied and SHOULD not be explicitly encoded in the `Hello` payload. A future implementation of OTLP that wants to use a certain optional capability SHOULD use `Hello` handshake to announce its capabilities and discover capabilities of the peer and adjust its behavior to match the expectation of a peer that does not support a particular capability.
```

Comment:
I think we should do it differently. If we do want to do a major change we can simply have an ExportV2 method on the rpc service.

Reply:
I thought about using version numbers in the protocol (what you suggest with ExportV2 is having version numbers, per service call or per protocol). I decided against it for the following reason:

In the future we will likely decide to extend OTLP with various new capabilities. Some of this capabilities will probably be optional, not required for all implementations. So, imagine that we have N optional capabilities and clients and server can implement a subset of this N capabilities. It will potentially we have 2^N versions of the protocol that can exist in the wild which need to be able to talk with each other. It would not be scalable to attempt to label these protocols by versions numbers.

This is the reason I decided to go with the concept of "capabilities" instead of version numbers. When a new capability is added to the protocol, it will be the responsibility of client and server to exchange the capabilities they support during `Hello` handshake.